neps/nep-0330.md

```diff
@@ -58,35 +71,42 @@ function contract_source_metadata(): ContractSourceMetadata {}
 
 ## Reference Implementation
 
-As an example, say there was an NFT contract deployed on-chain which was currently using the commit hash `39f2d2646f2f60e18ab53337501370dc02a5661c` and had its open source code located at `https://github.com/near-examples/nft-tutorial`. This contract would then include a struct which has the following fields:
+As an example, consider an NFT contract located at the relative path `nft-contract`, which was deployed using the `cargo near deploy --no-abi` environment docker image `sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471`. Its latest commit hash is `39f2d2646f2f60e18ab53337501370dc02a5661c`, and its open-source code can be found at `https://github.com/near-examples/nft-tutorial`. This contract would then include a struct with the following fields:
```

Comment:
@Canvinus @frol @robert-zaremba  Could we suggest having a `Cargo.lock` in the contract folder?

I think this way: if the contract was hacked because of a security issue in a 3rd-party dependency, then such an issue may not be reproducible after the dependency upgrade (in the least "patch" version triplet).

Also, I checked the `nft-tutorial/nft-contract`, it has `[patch.crates-io]` which leads to a custom crate, and it may be hard to reproduce the contract with such a dependency.

Reply:
Yes, good idea!

Reply:
Added note about Cargo.lock in 222d2602059765f1a127223840a72f2de64bb219
And changed the example repo to https://github.com/near/cargo-near-new-project-template in 35bed0fbc75783132b7ca95a11c9c51974844824

Reply:
@Canvinus Thanks but Сargo.lock is still necessary because dependencies of dependencies may change.

Reply:
@fadeevab, I have info about this in the comment of source_code_snapshot, where do you think it is also necessary to include information about Cargo.lock

Reply:
@Canvinus
Thank you for your clarification questions!

1. I see that https://github.com/near/core-contracts have `Cargo.lock`s, it could be used as an example.
2. I realize we already use  https://github.com/near/cargo-near-new-project-template: is it possible to commit `Cargo.lock` there?
3. I would suggest moving that sentence from the code comment to the specification (somewhere before the reference implementation).

```
### Cargo.lock

It is important to have `Cargo.lock` inside the source code snapshot to ensure reproducibility. Example: https://github.com/near/core-contracts.
```
👇
![image](https://github.com/near/NEPs/assets/5967447/815f3b2e-b511-47d7-b35b-a23ea664befc)

Reply:
Great that most of this project is still compiling, thanks to Cargo.lock: https://github.com/near/cargo-near-new-project-template. But I thought, isn't that a problem that they are using an outdated SDK that is not even capable of having contract source metadata? Wouldn't this be confusing to use as a reference?

Reply:
Added new section `Ensuring WASM Reproducibility` c06b054f4ec19e77ce92df2fe30a4bab5b8e0ea7 and moved info out of comments into it

Reply:
> Added new section `Ensuring WASM Reproducibility` [c06b054](https://github.com/near/NEPs/commit/c06b054f4ec19e77ce92df2fe30a4bab5b8e0ea7) and moved info out of comments into it

Awesome!

>  Wouldn't this be confusing to use as a reference?

An alternative is to add `Cargo.lock` to https://github.com/near/cargo-near-new-project-template and use it as a reference. It would be the best option probably...